### ISSUE TITLE
Kerberos

### ISSUE THREAD

**Issue by abn0mad** [kanidm/kanidm#1614]
**Is your feature request related to a problem? Please describe.**

No problem, feature suggestion / request / discussion opener

**Describe the solution you'd like**

Kerberos' implementation in Kanidm, primarily for the purposes of secure NFSv4 user home directories

**Describe alternatives you've considered**

Manually setting up a Kerberos installation and connecting it to Kanidm's LDAP implementation; haven't tried it yet.

**Additional context**

If Kanidm could at some point implement Kerberos, it would truly become a one-stop-shop for identity management. 

**Personal note**

My admiration, and thanks for having started the Kanidm project, fantastic work. When the admin web UI and clustering capability have been added, then Kanidm will be the perfect vehicle for FOSS based organisational networks. Very exciting and a sorely needed solution for less tech-savvy teams, and a genuine time saver in general. Maintaining an OpenLDAP + Kerberos + Keycloak (or Authelia) + NFS + SAMBA deployment is a real pain. Kanidm is well on the way to becoming the answer.

For “regular” trust environments, Kanidm is already close to being the perfect solution. More restrictive environments, on the other hand, such as those with the need to protect extra sensitive data (i.e. NGOs, activists, journalists, …) would greatly benefit from Kerberos integration to set up zero-trust environments. What I mean by that is environments where devices have no local storage, where users are not allowed to access USB storage devices for risk of leaking sensitive data etc.

This issue is intended to get the discussion started, I fully appreciate that this would be a tremendous undertaking and put a significant burden on the project. It would be nice if it was added somewhere down the line; either integrated directly in Kanidm, or by allowing the LDAP schema to be tweaked for an external Kerberos deployment (if necessary, haven't tried it yet, will try it soon).

Thanks for reading and above all, thanks for all the hard work in creating this wonderful project. My hat is off to you all.

**Comment by Firstyear** [comment]
Thanks for the very kind words @abn0mad. I'm really glad you like the project! 

Let's seperate out two different parts of your request here.

* Addition of Kerberos
* Ability to use Kanidm for authentication of Samba/NFS

It so happens that today, to authenticate Samba/NFS that requires Kerberos, which is why you are asking for this feature. 

So lets focus on that first. Today Samba can authenticate with Kerberos or NTLM. If we pursued NTLM, we need to expose passwords in a certain way to fileservers. There would be a lot of limits for windows file servers and services, but windows clients wouldn't have a problem. The hard part would be things like automounting home drives on user login or such. As well, by being limited to passwords, we need people to enter this password, or some other way to have the password exposed to the correct parts of the stack. Generally it gets pretty messy pretty quick, needing to shuffle around shared secrets. 

Samba can do kerberos for authentication, and this would be better in a way, since when you login you get a credentials cache - we wouldn't need to actually use the AS_REQ/kinit here, we could adapt kanidm unixd to have a custom method to initialise the ccache which would allow us to actually use 2FA. Windows still would expose a username password dialog though, so we're back to needing a password that the user can enter. Kerberos on a non-unixd client would also likely need a password to kinit for AS_REQ or PKINIT - or even something custom. Again, this gets hugely messy. Especially if we have ssh authentication because then post login you need to kinit somehow, but you'd already have logged in so you couldn't mount your home share .... 

Anyway it's all a bit of a mess. NFS has pretty much the same limitations, where it's authentication is IP bound or Kerberos only.

For server to server authentication though, you don't even need to involve kanidm. You would just have host keytabs on each machine and then this is what authenticates the machine to the fileserver. You could consider this instead, since for kanidm you login to the host machine, then the host machine authenticates to the file server on your behalf. This would already be posible, no kanidm involvement needed. 


But the better option - and one I'll be discussing with some people in the next two weeks - is a way to authenticate to Samba / NFS that *doesnt* need Kerberos. This would actually be better since then we don't need Kerberos at all, and we could use something that is better suited to how Kanidm works.

To summarise

* Host to Host (server to server) authentication today with kerberos for fileshares is possible, and needs no involvement from Kanidm.
* We want to try to have Samba or NFS develop alternate authentication methods both for server to server and user to server authentication
* Kerberos in Kanidm would always have strict limits on how it works, since Kerberos and it's concepts tend to be invasive and taint the world. We don't want Kerberos to be the centre of our system, but an artefact that's just for certain server to server or user to server contexts in very limited capacity.

**Comment by abn0mad** [comment]
@Firstyear thanks for the very thorough and understandable reply, much appreciated :)

Indeed, it is because of NFS and Samba that I was asking (although primarily NFS + Linux clients with auto-mounting home directories). If the Kanidm team can achieve NFS / SAMBA authentication without needing Kerberos, then that would be an even better solution indeed. I had no idea that was an option. That would be fantastic.

With the upcoming Admin UI, clustering, replication and NFS/SMB authentication, Kanidm would truly be a gift beyond imagining. Finally, an easy-to-use alternative to Active Directory that doesn't require major tinkering and maintenance.

Thank you so much for reading, answering and taking it into consideration – along with all the hard work on the Kanidm in general. 

It is my intention to include Kanidm in some tutorials and guides on setting up FOSS-based networks and servers; I hope it can motivate people and organisations to become sponsors of this so well-deserving project.

**Comment by Firstyear** [comment]
After discussion at samba xp, it seems like the future is "kerberos" for smb auth, so we likely need to consider a design for this - but I don't think it will be a traditional kinit style design, since we need ticket issuance for users when they login to a machine via ssh. This means we likely need a way to have a machine to request and fetch a ticket on behalf of a user through some form of machine authorisation/authentication. 

As an alternative, it was pointed out by @scabrero that as part of #200 it may be possible to add a custom SPNEGO module to allow non-kerberos auth on windows clients. 

I think some more design and discussion about the use cases here will be needed to make sure we actually design and implement something that is reliable and works well, but I also am wary that we don't want to implement all of kerberos (like trusts) since the goal is just for user-to-storage auth or server-to-storage auth.

**Comment by Firstyear** [comment]
There may be an alternative with: https://medium.com/@mor2464/azure-ad-pass-the-certificate-d0c5de624597

**Comment by Firstyear** [comment]
We are currently discussing with other projects about how to proceed here, but most likely we will pursue the PKU2U implementation since this gives us a kerberos client to server implementation without the normal risks/baggage of a full KDC.

**Comment by Firstyear** [comment]
> > We are currently discussing with other projects about how to proceed here, but most likely we will pursue the PKU2U implementation since this gives us a kerberos client to server implementation without the normal risks/baggage of a full KDC.
> 
> One of the hangups I've had with `kanidm` was how to handle shared user home directories across systems that I've traditionally handled with FreeIPA and NFSv4. PKU2U seems like a good solution, but it doesn't appear that Samba supports it at all, so would this be expected to only be for Windows SMB?

Something that isn't visible in this ticket is that I was recently at an in person meeting with members of the Samba project, and they are interested to pursue PKU2U support in the project. We will be working with them to integrate that functionality to Kanidm as well. So this is a bigger project than just Kani - the Samba team members also want Kanidm to have a good smb auth solution too, so this will happen! :) 

> 
> Maybe this is actually potentially a separate issue on integrating network shares vs this specific open issue, but I have yet to come up with a good model on how to handle user mounts for filesystems that isn't based purely off of NFSv3 (unsafe) or using some rather hackish SSHFS automation with systemd. Have you encountered anyone else coming up with good solutions for this @Firstyear ?

At this time, no sorry. You can do host-to-host shares with nfs3/4 based on ip security, and then rely on the hosts to protect the user security boundaries, but we haven't actually done anything with mounting on login yet. Likely we would be able to extend the tasks daemon with this functionality to mount on login when we also have PKU2U. In addition the PKU2U changes would likely also allow nfs4 with krb to function too since a lot of the changes have to happen in MIT krb IIRC. 

The reason we haven't done mount-on-login yet is simply that we don't have a good way to provision and handle the authentication details. We could hack something with Samba to distribute ntlm hashes into smbd and then have them exposed to unixd via a sucessful unix auth token response, but it wouldn't work on ssh login because we don't have a password exchange. Where as with pku2u we are using the machines host certificate as a security manager to say "yes I swear this person just logged in pls mount their home".

**Comment by Firstyear** [comment]
@dmulder mentioned this crate which may help us for storing credentials into a ccache. 

https://crates.io/crates/kerberos_ccache

**Comment by Firstyear** [comment]
UPDATE: Where are we at here?

Recently @dmulder has been looking into how to inject an AS_REP into a ccache so that we can create these from domain joined machines.

The "vision" is that authentication to a machine should grant you the CCACHE without a direct kinit or use of a password. 

So in the mean time what do we want to do?

* Make rust-crypto der based parsers for krb rfc types - currently we were using picky-asn1 but the authors of that library prefer the rc-der types.
* Implement CS3 (AKA CTS) mode for rust-crypto over block modes. This is because krb uses AES-128-CTS instead of anything sane like CBC+PKCS7 or GCM. 
* Implement a minimal kinit client based on the above to test the "client side" of decryption of an ASREP. 
* Once we have that, we can then test writing the server side since we know our client is valid. 
* Investigate what's needed for the server side for client to server auth.

**Comment by 2xsaiko** [comment]
Something that hasn't been mentioned yet: I just thought about how Windows lets you log in via PIV smartcards, and this of course works with AD/Kerberos too. It seems like MIT krb5 supports them as well, and a yubikey can apparently act as a smartcard. If you can register a smartcard in Kanidm, could this be another way to have passwordless login for Kerberos (and maybe even for Kanidm in general)?

**Comment by Firstyear** [comment]
Yes, this was also planned :)

**Comment by TheDragon44** [comment]
Was curious on two points - what's the state of play on this? 

Also (apologies slightly off topic) is there any method to donate to the project? Quick search of issues and around the repo/README/manual but couldn't see it mentioned anywhere obvious.

Thanks for all you're doing here btw, I echo the sentiments of the original author of this issue - even if I'm still at the implementation stage... 

The docs are **so** thorough, thank you for all the time you invest in this project 🙂

**Comment by Firstyear** [comment]
> Was curious on two points - what's the state of play on this?

https://github.com/kanidm/libkrimes/

We have a working KDC, and I will soon be visiting the collaborator on this for a hack session to get it working with windows/linux clients, as well as extending the support for SMB attributes needed. 

> 
> Also (apologies slightly off topic) is there any method to donate to the project? Quick search of issues and around the repo/README/manual but couldn't see it mentioned anywhere obvious.

At this time we don't have a method to accept donations. 

> 
> Thanks for all you're doing here btw, I echo the sentiments of the original author of this issue - even if I'm still at the implementation stage... the docs are **so amazingly thorough**, thank you for all the time you invest in this project 🙂

No problem, I'm so glad that you appreciate it :)

**Comment by Gabgobie** [comment]
Hi,

thank you for this awesome project. I've been taking a look at libkrimes (or rather the README thereof) and watching this issue for a while. Kerberos has been a major blocker for me to migrate.

I don't want to waste your time with just a comment "+1ing" the feature request but instead I'd like to get an understanding on the scope you envision for the kerberos integration in Kanidm.

The README of libkrimes mentions intentional limitations as to not compromise on security. These limitations are yet to be decided. Do these limitations refer to behavior such as limiting the available level of encryption or will they reach into limiting compatibility with specific services?

In this issue, everyone has been talking about KRB5 as a means to integrate with storage providers and while I can see this being one of the main use-cases, I think Kerberos is currently needed in other areas as well. To name two non-storage examples in my own environment:

- Mail
  - Unless this document is outdated, [Thunderbird supports OAuth only for certain hardcoded email providers](https://wiki.mozilla.org/Thunderbird:Autoconfiguration:ConfigFileFormat#OAuth2), meaning I still need to rely on KRB5 authentication to my internal mail server. They seem to have been chewing on this for the past 6 years https://bugzilla.mozilla.org/show_bug.cgi?id=1602166
- Git(Lab)
  - Not having to deal with SSH keys or access tokens has been pure bliss. To my understanding, git does not officially support using OAuth tokens directly.

Best,
Gab

**Comment by Firstyear** [comment]
> 
> The README of libkrimes mentions intentional limitations as to not compromise on security. These limitations are yet to be decided. Do these limitations refer to behavior such as limiting the available level of encryption or will they reach into limiting compatibility with specific services?

Both - we will only offer secure krb modes, and we will only offer specific service integrations. A challenge is krb enshrines that a user only has a password, and other methods to get a ticket like pkinit require a full smartcard deployment. 

> 
> In this issue, everyone has been talking about KRB5 as a means to integrate with storage providers and while I can see this being one of the main use-cases, I think Kerberos is currently needed in other areas as well. To name two non-storage examples in my own environment:
> 
>     * Mail
>       
>       * Unless this document is outdated, [Thunderbird supports OAuth only for certain hardcoded email providers](https://wiki.mozilla.org/Thunderbird:Autoconfiguration:ConfigFileFormat#OAuth2), meaning I still need to rely on KRB5 authentication to my internal mail server. They seem to have been chewing on this for the past 6 years https://bugzilla.mozilla.org/show_bug.cgi?id=1602166

This reduces your mail security to a single krb password, and the same krb password as you access storage with. I dont like this as an idea, I think oauth2 would be better, but application passwords that we have in development is better. 

> 
>     * Git(Lab)
>       
>       * Not having to deal with SSH keys or access tokens has been pure bliss. To my understanding, git does not officially support using OAuth tokens directly.

It doesn't but you can probably find a way to pull ssh keys from kanidm into gitlab, which would automatically populate these.

Both of these aren't cases where I'd want krb involved, krb really should be limited to what is absolutely necessary, where we have no other choice.

**Comment by Gabgobie** [comment]
Thanks for the explanation.

This leaves me with some confusion. To my understanding you want to avoid using passwords at any cost leaving pkinit as the only available alternative. So far, I've been using FreeIPA in my environment and successfully deployed it without the use of passwords via Passkeys (USB-Stick).

I never implemented kerberos myself but only set it up in my environment. That being said, unless my understanding of kerberos is fundamentally flawed, I also don't quite get what you mean with the single password for all services unless you are referring to the authentication to the identity provider. To my understanding KRB is using the password once to get the TGT which is then used as a "short-lived" secret to get service tickets which again are short lived and each with a narrow scope. This same limitation should apply to OAuth, where you need to authenticate with Kanidm using the user's secret, be it a Passkey or anything else to then authenticate to each service using OAuth.

Note that I don't want to derail the topic so feel free to abort this conversation at any point.

Best,
Gab

> Edit: I decided to edit my most recent comment on this thread to avoid pinging everyone. Maybe it's helpful to somebody.
> 
> > Just silently wanted to drop this here: https://freeipa.readthedocs.io/en/latest/workshop/12-external-idp-support.html
> > 
> > Until krb5 is ready in Kani, proxying with FreeIPA may be a stopgap measure for some scenarios.
> > 
> > Disclaimer: I haven't implemented it in my environment yet, just came across it in the docs. I did not look into operational and security implications.

**Comment by dblsaiko** [comment]
> A challenge is krb enshrines that a user only has a password, and other methods to get a ticket like pkinit require a full smartcard deployment.

Did you come to a conclusion on having no traditional kinit at all, where instead this is handled by some Kanidm-specific software (Kanidm client or unixd), that authorizes via existing secure auth methods in Kanidm and then adds the TGT into the credentials cache, like you described in previous comments? Or what you described in https://github.com/kanidm/kanidm/issues/1614#issuecomment-2111841849 (can't tell if that is the same approach, I'm not deep enough in the Kerberos sauce to know what these terms mean :P) That would mean no dependency on a password, as far as I can tell, since that is only used for the basic kinit.

I assume the problem there would at least be compatibility, since it would no longer work with just a standard Kerberos client.

**Comment by Firstyear** [comment]
> Thanks for the explanation.
> 

> I never implemented kerberos myself but only set it up in my environment. That being said, unless my understanding of kerberos is fundamentally flawed, I also don't quite get what you mean with the single password for all services unless you are referring to the authentication to the identity provider. To my understanding KRB is using the password once to get the TGT which is then used as a "short-lived" secret to get service tickets which again are short lived and each with a narrow scope. This same limitation should apply to OAuth, where you need to authenticate with Kanidm using the user's secret, be it a Passkey or anything else to then authenticate to each service using OAuth.

Yes, that's the issue - getting the TGT is the *single* password. And the mechanisms that are universally accepted by all platforms for that all have viable offline attacks.  (We won't when we deploy krb though, as we plan to use KKDCP which is KRB over HTTPS). That TGT can request as many service tickets, to whatever services it wants as there is no scoping of the TGT. Once you have it, you can access everything. 

Where as in OAuth/OIDC you can think of it as a "TGT per service" if you want to make a horrific analogy. This way, each credential is scoped and can have unique permissions or limits. Compromise of one, doesn't compromise everything. And we protect. Even in Kanidm, your session token internally has scopes that can be activated/deactivated so that theft of a token can do less damage. 

Anyway, that's why when we do KRB we want to make it so that it's bound to a single machine if possible, and that it's a strong high entropy password, and only for "as few services as possible". 

Remember the other part of this is that kerberos is a nightmare to debug, and that's generally what also limits it's usage is that barrier to integration and configuration. 

TGT's aren't shortlived by the way. They are able to survive for days in most default configurations. Contrast to minutes for OAuth2. 

> Did you come to a conclusion on having no traditional kinit at all, where instead this is handled by some Kanidm-specific software (Kanidm client or unixd), that authorizes via existing secure auth methods in Kanidm and then adds the TGT into the credentials cache, like you described in previous comments? Or what you described in [#1614 (comment)](https://github.com/kanidm/kanidm/issues/1614#issuecomment-2111841849) (can't tell if that is the same approach, I'm not deep enough in the Kerberos sauce to know what these terms mean :P) That would mean no dependency on a password, as far as I can tell, since that is only used for the basic kinit.

We did think of this! The problem is how to get that client software onto machines. Like a windows machine. Or a mac. Or a printer. Or an apple tv etc. 

The big user and reason we are looking at kerberos at all is SMB fileshares. Microsoft is attempting to remove NTLM (which is good) but they are going to make kerberos the only option (Bad, IMO). But we have to follow what they are doing, so that means we need to have kerberos for SMB.

Now think about SMB clients - we can't install something on an iPhone or Android to do a kinit. We can't assume they'll have PKINIT setup (smartcards). And while there are some thought about ctap/webauthn init for credentials, we can't guarantee that anything will have that, nor can you connect a security key to many devices. So what's the only universal accepted mechanism? Password. That means the strongest we can assume kerberos to be, is it's weakest link. 

So we can take a lot of steps to try to make that difficult to attack - the libkrimes kdc for example is the only KDC that actually increases the hashing rounds on preauth for example. We also looked at cuda hashcat to see what level of rounds and password minimums we can use to defend against preauth attacks. Finally we want to use KKDCP to protect the pre-auth/transit of that vulnerable material. 

But at the end of the day, we still are reliant on a password. That's single factor, so we shouldn't have krb used as a single factor, to access services that should (or must) have MFA on them. That includes email, ssh etc. 

Now, we could do tricks to make a TGT only after a trusted machine login etc, and somehow mark that TGT as special and usable with those services, but in the initial work, I don't want to do that. SMB fileshares are the goal, so that's our focus right now,

**Comment by neverreadsemail** [comment]
I’ve several times here seen PKINIT conflated with smartcard infrastructure. I want to clarify that PKINIT, the krb5 protocol extension is about getting TGTs given an X.509 certificate, and is not specific to smartcards per se. that said, you’re still back to how do you get an X.509 certificate signed by a CA the kdc trusts and which matches the KDC’s authentication policy onto the SMB client. And then get the client to use that to acquire a krb5 TGT.

**Comment by neverreadsemail** [comment]
> TGT's aren't shortlived by the way. They are able to survive for days in most default configurations. Contrast to minutes for OAuth2

my understanding is that OAuth access tokens are short lived(minutes) but that refresh tokens are more comparable to TGTs and have similar lifetimes(hours to days). How is this different?(I’m much more familiar with krb5(have admined various realms since the mid 90s) than with oauth, so this is genuine curiosity, not trying to contradict.)

also, last I checked, the default for MIT krb5 is 10 hours. Maybe windows KDCs default to something longer? I get that’s still fairly long, but still generally what most consider short lived. IME “long lived” is typically used to talk about things like HTTPS server certs or a krb5 instance key which last months, years, or even “forever”(until revoked)

**Comment by Firstyear** [comment]
> I’ve several times here seen PKINIT conflated with smartcard infrastructure. 

PKINIT is an option but it's obscenely complex.  We don't want to centre kerberos for auth at the end of the day, it's simply too hard and too complex to integrate even to browsers, webservers etc. 

> my understanding is that OAuth access tokens are short lived(minutes) but that refresh tokens are more comparable to TGTs and have similar lifetimes(hours to days). How is this different?(I’m much more familiar with krb5(have admined various realms since the mid 90s) than with oauth, so this is genuine curiosity, not trying to contradict.)

They can't be compared directly at all. They are not quite the same.

OAuth2 tokens have a 1 to 1 ratio between the server and the oauth2 client. So access tokens from "nextcloud" can't work with "imap".

But in KRB, the TGT allows you to request TGS for any service. One TGT lets you access nextcloud, imap etc. So suddenly the ability to move laterally from one TGT is such a bigger issue.

Second, kerberos services absolutely break if you use short lived anything. I tried with 5 minute long TGS times and it trashed everything. 

Where as in OAuth I can happily have a 5 minute access token, and nothing explodes. 

That's what's important here. OAuth2 with 5 minute access token lives *forces* the refresh token to be use frequently and often which can allow revocation of that token to occur. 

But with KRB once you have the TGT + TGS, that's it. It's valid for 8 more hours. 

If I lock an account in Kanidm, that users oauth2 sessions will stop working within minutes. For kerberos? It's hours.

That's why it's so much more dangerous. 

> 
> also, last I checked, the default for MIT krb5 is 10 hours. Maybe windows KDCs default to something longer? I get that’s still fairly long, but still generally what most consider short lived. IME “long lived” is typically used to talk about things like HTTPS server certs or a krb5 instance key which last months, years, or even “forever”(until revoked)

Remember, the important factors here is likelihood of compromise and the consequences of that. Certificates have revocation lists that can be revoced quickly, oauth2 can have short token lives. KRB doesn't get any of these toys and so you have a case where an attacker can steal a TGT and run rampant for hours.
